Asymmetric line flow (Refined); READY FOR REVIEW !!!

sambuddhac opened the pull request:
## Description


This PR attempts to introduce bidirectional asymmetric transmission lines into the system. Such lines have different values of MW flow limits, loss percentage, and capacity expansion limits along the two directions. While doing so, this PR also ensures that there are also symmetric lines present in the system along with the asymmetric lines and modifies the code in such a way that backwards compatibility is maintained with the existing cases.


## What type of PR is this? (check all applicable)

- [x] Feature
- [ ] Bug Fix
- [x] Documentation Update
- [ ] Code Refactor
- [ ] Performance Improvements

## Related Tickets & Documents
<!-- 
Please use this format to link issue numbers: Fixes #123
https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword 
-->

## Checklist

- [x] Code changes are sufficiently documented; i.e. new functions contain docstrings and .md files under /docs/src have been updated if necessary.
- [x] The latest changes on the target branch have been incorporated, so that any conflicts are taken care of before merging. This can be accomplished either by merging in the target branch (e.g. 'git merge develop') or by rebasing on top of the target branch (e.g. 'git rebase develop'). Please do not hesitate to reach out to the GenX development team if you need help with this.
- [x] Code has been tested to ensure all functionality works as intended.
- [x] CHANGELOG.md has been updated (if this is a 'notable' change).
- [x] I consent to the release of this PR's code under the GNU General Public license.

## How this can be tested

<!--
If applicable: What cases should we try before/after? Will this alter any outputs, or is it a strictly internal change?
-->

<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.
  
  For a timely review/response, please avoid force-pushing additional
  commits if your PR already received reviews or comments.
  After the PR is approved we will give you a chance to tidy up the branch before merging.
  
  Before submitting a Pull Request, please ensure you've done the following:
  - 👷‍♀️ Create small PRs. In most cases, this will be possible. In the case of large feature or module additions, it is best to work towards a "minimum viable product" that is thoroughly tested but may not have all desired functionality or features, make a PR for this, and then later work to add more features.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation.

-->

## Post-approval checklist for GenX core developers
After the PR is approved

- [x] Check that the latest changes on the target branch are incorporated, either via merge or rebase
- [x] Remember to squash and merge if incorporating into develop

sambuddhac committed:
Refined Asymmetric Bidirectional Flow

sambuddhac committed:
Modified the line flow limits with asymmetric bidirectional flow limits

sambuddhac committed:
Updated files for investment_transmission and load_network

sambuddhac committed:
Update transmission.jl

sambuddhac committed:
Modified the transmission write output files to reflect bidirectional asymmetrical flows

sambuddhac committed:
Modified load_network_data.jl and Network.csv to include additional columns for bidirectional asymmetric flow

sambuddhac committed:
Got rid of unnecessary poitive nd negative reinforcement cost

sambuddhac committed:
Corrected the cost term for bidirectional investment

sambuddhac committed:
Refined asymmetric flows; Debugging

sambuddhac committed:
Write output files and doc pages for investment transmission updated

sambuddhac committed:
Write output files and doc pages for investment transmission updated

sambuddhac committed:
Added new example case with zero capacities in forward direction and refined write_expansion

sambuddhac committed:
Added new example case

sambuddhac committed:
Resolved bidirectional flow result bug on asymmetric lines by explicit partitioning of symmetric and asymmetric rows on network dataframe and introducing additional associated decision variables into the code-base

sambuddhac committed:
Revised loss reformulation and added benchmark test cases for validating bidirectional asymmetric flows

sambuddhac committed:
precompile bug fixed

sambuddhac committed:
Corrected and refined quadratic loss reformulation

sambuddhac commented:
Hi @JesseJenkins and @lbonaldo , I have pushed all the modifications to the code base for this PR. Few things to note: First of all, as you will see, it's way more lengthy and complicated than what Jesse suggested to me previously. I first tried with the simpler version. But, was getting bidirectional flows on the asymmetric lines with Qian's example case even when we restricted the flow limits to zero in one of the directions for the asymmetric lines. With some pondering, I realized that I was doing wrong indexing on the constraints and decision variables. So, I had to do a bit more digging and refining the code. The fundamental reasons for which the code is so much lengthier are: 1) I have both symmetric as well as asymmetric lines in a system and 2) In order to maintain backward compatibility with the existing example cases (As you will observe, I had to go to lengths to split the data-frame into two parts and assign separate variables and constraint names so that things stay clean and correct. We tested with @qluo0320github 's example cases and it seems the results are sensible (?) However, I still get discrepancies when I benchmark against an actually symmetric system versus a "simulated" symmetric system (in which the asymmetric lines have the same flow limits, loss percentages, and transmission buildout in both directions). I am attaching hereby all the relevant flow comparison 
[Flow_Comparisons.xlsx](https://github.com/user-attachments/files/18159896/Flow_Comparisons.xlsx)
result file (the names should be self-explanatory) here for you to take a look. I scrutinized the code several times, but couldn't spot any flaw so far. I would appreciate if you could point some obvious bug (if at all) that I might have missed. Also, I would appreciate taking a look at the quadratic loss formulation. I believe I did it right, but am not totally certain. Thank you so much !!! I have also updated most of the doc pages; I will wrap up parts of the transmission.jl doc page and make another push to close this.

sambuddhac committed:
Doc update and fixed absolute value of flow constraint for asymmetric lines

sambuddhac committed:
Modified CHANGELOG

sambuddhac committed:
Merge conflict bug fix

sambuddhac committed:
Fixed absolute value of flow RHS constraint by imposing tighter bounds

sambuddhac committed:
Corrected indexing

lbonaldo commented on CHANGELOG.md:
```diff
@@ -25,6 +25,9 @@ settings file (PR #801).
 charge and storage variables (#760 and #763).
 - Deduplicated docs on optimized scheduled maintenance for thermal resources (#745).
 - Removed the `CapRes_*` columns from `Network.csv` since they were not being used (#784).
+- Modified `transmission.jl`, `investment_transmission.jl`, `load_network_data.jl`, 
+`write_transmission_flows.jl`, `write_transmision_losses.jl`, and `write_network_expansion.jl`
+for implementing asymmetric bidirectional flows. (#789)
```
This part should go under the `##Unreleased` section above as it belongs to a new release.

lbonaldo commented on example_systems/1_three_zones_asymmetric_bidirectional/README.md:
```diff
@@ -0,0 +1,32 @@
+# Three Zones
```
If we decide to keep this example, we might need to this description.

sambuddhac replied:
Agreed !!!

lbonaldo replied:
have we decided to keep this example?

lbonaldo commented on example_systems/1_three_zones_asymmetric_bidirectional/settings/genx_settings.yml:
```diff
@@ -0,0 +1,14 @@
+UCommit: 2 # Unit committment of thermal power plants; 0 = not active; 1 = active using integer clestering; 2 = active using linearized clustering
+TimeDomainReduction: 1 # Time domain reduce (i.e. cluster) inputs based on Demand_data.csv, Generators_variability.csv, and Fuels_data.csv; 0 = not active (use input data as provided); 0 = active (cluster input data, or use data that has already been clustered)
+OutputFullTimeSeries: 1
+asymmetrical_trans_flow_limit: 1 # Switch to indicate if asymmetrical bidirectional lines are included in the system or not
```
By convention, we use PascalCase (or UpperCamelCase) for setting keys: `AsymmetricalTransFlowLimit`.

sambuddhac replied:
Done !!!

lbonaldo commented on example_systems/1_three_zones_asymmetric_bidirectional/system/Network.csv:
```diff
@@ -0,0 +1,6 @@
+﻿,Network_zones,Network_Lines,Start_Zone,End_Zone,Asymmetrical,Line_Max_Flow_MW,Line_Max_Flow_MW_Neg,transmission_path_name,distance_mile,Line_Loss_Percentage,Line_Loss_Percentage_Neg,Line_Max_Reinforcement_MW,Line_Max_Reinforcement_MW_Neg,Line_Reinforcement_Cost_per_MWyr,DerateCapRes_1,CapRes_1,CapRes_Excl_1
+MA,z1,1,1,2,0,2950,2950,MA_to_CT,123.0584,0.012305837,0.012305837,2950,2950,12060,0.95,0,0
+CT,z2,2,1,3,0,2000,2000,MA_to_ME,196.5385,0.019653847,0.019653847,2000,2000,19261,0.95,0,0
+ME,z3,4,1,2,1,2000,4000,MA_to_CT,123.0584,0.012305837,0.12305837,1206,2412,12075,0.95,0,0
+,,5,1,3,1,1000,800,MA_to_ME,196.5385,0.019653847,0.19653847,1926,963,19200,0.95,0,0
+,,6,2,3,1,500,1200,CT_to_ME,196.5385,0.019653847,0.19653847,1206,603,12000,0.95,0,0
```
I believe the `Network_Lines` column should be `[1, 2, 3, 4, 5]` here. I'm getting an error when trying to run this case.

lbonaldo commented on example_systems/6_three_zones_w_multistage_asymmetric_bidirectional/inputs/inputs_p3/system/Network.csv:
I'm encountering the following error message when trying to run this example:
```
Reading Input CSV Files
ERROR: ArgumentError: column name :Asymmetrical not found in the data frame
```

sambuddhac replied:
Just fixed it with introducing a new example case for running multistage with asymmetric flows.

sambuddhac replied:
@lbonaldo  I fixed this with the latest push and tested also. Introduced a new example case for this.

lbonaldo commented on src/configure_settings/configure_settings.jl:
```diff
@@ -37,7 +37,9 @@ function default_settings()
         "PoliciesFolder" => "policies",
-        "ObjScale" => 1)
+        "ObjScale" => 1,
+        "asymmetrical_trans_flow_limit" => 0
```
Same as above, we typically use PascalCase (or UpperCamelCase) for setting keys.

lbonaldo commented on src/load_inputs/load_network_data.jl:
```diff
@@ -15,24 +15,53 @@ function load_network_data!(setup::Dict, path::AbstractString, inputs_nw::Dict)
     L = length(as_vector(:Network_Lines))
     inputs_nw["L"] = L
+    L_asym  = 0 # Default number of asymmetrical lines
+    if setup["asymmetrical_trans_flow_limit"] == 1
+        L_asym = length(filtered_vector(network_var, :Asymmetrical, 1, :Network_Lines)) #Number of asymmetrical lines
+    end
```
This line will throw an error if `["asymmetrical_trans_flow_limit"] == 1` but the `:Asymmetrical` column is not found in the `Network.csv` file. We could add a validation to ensure that:
- If the user sets ["asymmetrical_trans_flow_limit"] == 1 
- Then a corresponding set of additional columns **must be** provided.

sambuddhac replied:
@lbonaldo  I fixed this with the latest push and tested also. In order to review, for instance, you could take any of the usual example cases (i.e. not the asymmetrical cases) and in the settings file, put the AsymmetricalTransFlowLimit switch and set it to 0

sambuddhac replied:
Fixed this now.

lbonaldo commented on src/write_outputs/transmission/write_nw_expansion.jl:
```diff
@@ -1,12 +1,32 @@
+    for i in intersect(SYMMETRIC_LINE_INDEX, EXPANSION_LINES)
+        transcap[i] = value.(EP[:vNEW_TRANS_CAP][i])
+    end
+    for i in EXPANSION_LINES_ASYM
+        transcap_pos[i] = value.(EP[:vNEW_TRANS_CAP_Pos][i])
+        transcap_neg[i] = value.(EP[:vNEW_TRANS_CAP_Neg][i])    
```
```suggestion
   for i in eachindex(EXPANSION_LINES_ASYM)
        asym_line_index = EXPANSION_LINES_ASYM[i]
        transcap_pos[i] = value.(EP[:vNEW_TRANS_CAP_Pos][asym_line_index])
        transcap_neg[i] = value.(EP[:vNEW_TRANS_CAP_Neg][asym_line_index])    
```

lbonaldo commented on src/write_outputs/write_outputs.jl:
```diff
@@ -122,14 +122,14 @@ function write_outputs(EP::Model, path::AbstractString, setup::Dict, inputs::Dic
         end
 
-        if output_settings_d["WriteTransmissionLosses"]
+        #=if output_settings_d["WriteTransmissionLosses"]
```
Can we uncomment this part?

sambuddhac replied:
This is one of the bugs, @lbonaldo , that I need some help from you to debug. That's why I commented it for now.

sambuddhac replied:
@lbonaldo could you please take a look at this one. This is one bug I couldn't fix.

lbonaldo reviewed (changes requested):
First round of review. Thanks @sambuddhac for this PR. In your opinion, which example case should we keep? The branch is now rebased with the latest develop.

lbonaldo commented on src/write_outputs/transmission/write_nw_expansion.jl:
```diff
@@ -1,12 +1,32 @@
 function write_nw_expansion(path::AbstractString, inputs::Dict, setup::Dict, EP::Model)
-    L = inputs["L"]     # Number of transmission lines
+    # Number of lines in the network
+    L_sym = inputs["L_sym"]
+    L_asym = inputs["L_asym"] #Number of transmission lines with different capacities in two directions
```
These two variables do not seem to be used

lbonaldo commented on src/model/core/transmission/transmission.jl:
```diff
@@ -82,21 +112,34 @@ As with losses option 2, this segment-wise approximation of a quadratic loss fun
     EnergyShareRequirement = setup["EnergyShareRequirement"]
     IncludeLossesInESR = setup["IncludeLossesInESR"]
 
+    SYMMETRIC_LINE_INDEX = inputs["symmetric_line_index"]
+    ASYMMETRIC_LINE_INDEX = inputs["asymmetric_line_index"]
```
It seems that the convention for `inputs` keys is to use capitalized words separated by underscores. What do you think about applying this convention to these two keys as well?

sambuddhac replied:
@lbonaldo this is also now fixed.

lbonaldo commented on src/load_inputs/load_network_data.jl:
```diff
@@ -15,24 +15,53 @@ function load_network_data!(setup::Dict, path::AbstractString, inputs_nw::Dict)
 
+    if setup["asymmetrical_trans_flow_limit"] == 1
+        # Transmission capacity of the network for asymmetrical lines; return direction(in MW)
+        inputs_nw["pTrans_Max_Possible_Neg"] = inputs_nw["pTrans_Max_Neg"] = to_floats(:Line_Max_Flow_MW_Neg) / scale_factor  # convert to GW
```
Another small suggestion: we usually place the units at the end of the key, e.g., `Inv_Cost_Per_MWyr`, `Line_Max_Flow_MW, etc`

sambuddhac replied:
Will fix it once we decide upon a final list of example cases. Otherwise, I will have to modify a whole of cases. Is that fine?

lbonaldo commented on src/model/core/transmission/investment_transmission.jl:
```diff
@@ -28,54 +48,99 @@ The additional transmission capacity, $\bigtriangleup\varphi^{cap}_{l} $, is con
-    L = inputs["L"]     # Number of transmission lines
+    
+    # Number of lines in the network
+    L_sym = inputs["L_sym"]
+    L_asym = inputs["L_asym"] #Number of transmission lines with different capacities in two directions
+    L = inputs["L"]
```
These three variables do not seem to be used.

lbonaldo commented on src/load_inputs/load_network_data.jl:
```diff
@@ -79,26 +112,34 @@ function load_network_data!(setup::Dict, path::AbstractString, inputs_nw::Dict)
         # Network lines and zones that are expandable have non-negative maximum reinforcement inputs
-        inputs_nw["EXPANSION_LINES"] = findall(inputs_nw["pMax_Line_Reinforcement"] .>= 0)
-        inputs_nw["NO_EXPANSION_LINES"] = findall(inputs_nw["pMax_Line_Reinforcement"] .< 0)
+        inputs_nw["EXPANSION_LINES"] = findall(inputs_nw["pMax_Line_Reinforcement"] .> 0)
```
I think we could do the intersection with `SYMMETRIC_LINE_INDEX` here to avoid needing to do it throughout the rest of the code.

sambuddhac replied:
Fixed with the latest push

lbonaldo commented on src/load_inputs/load_network_data.jl:
```diff
@@ -79,26 +112,34 @@ function load_network_data!(setup::Dict, path::AbstractString, inputs_nw::Dict)
 
     ## Sets and indices for transmission losses and expansion
     inputs_nw["TRANS_LOSS_SEGS"] = setup["Trans_Loss_Segments"] # Number of segments used in piecewise linear approximations quadratic loss functions
     inputs_nw["LOSS_LINES"] = findall(inputs_nw["pTrans_Loss_Coef"] .!= 0) # Lines for which loss coefficients apply (are non-zero);
```
Same as below, we could do the intersection with SYMMETRIC_LINE_INDEX here instead of throughout the rest of the code.

sambuddhac replied:
Fixed with the latest push

lbonaldo commented on src/model/core/transmission/transmission.jl:
```diff
@@ -82,21 +112,34 @@ As with losses option 2, this segment-wise approximation of a quadratic loss fun
+    LOSS_LINES_ASYM = inputs["LOSS_LINES_ASYM"] # Lines for which loss coefficients apply (are non-zero);
+    if NetworkExpansion == 1
+        # Network lines and zones that are expandable have non-negative maximum reinforcement inputs
+        EXPANSION_LINES = inputs["EXPANSION_LINES"]
```
These two variables do not seem to be used.

lbonaldo commented on src/model/core/transmission/transmission.jl:
```diff
@@ -82,21 +112,34 @@ As with losses option 2, this segment-wise approximation of a quadratic loss fun
+    if NetworkExpansion == 1
+        # Network lines and zones that are expandable have non-negative maximum reinforcement inputs
+        EXPANSION_LINES = inputs["EXPANSION_LINES"]
+        EXPANSION_LINES_ASYM = inputs["EXPANSION_LINES_ASYM"]
```
These two variables do not seem to be used.

sambuddhac committed:
Update CHANGELOG.md

sambuddhac committed:
Update CHANGELOG.md

sambuddhac committed:
Update genx_settings.yml

sambuddhac committed:
Update configure_settings.jl

sambuddhac committed:
Update src/write_outputs/transmission/write_nw_expansion.jl

Co-authored-by: Luca Bonaldo <39280783+lbonaldo@users.noreply.github.com>

sambuddhac committed:
Update write_nw_expansion.jl

Deleted L_sym

sambuddhac committed:
Update investment_transmission.jl

Deleted unused L_asym, L_sym, and L

sambuddhac committed:
Update Network.csv

Corrected line numbering

sambuddhac committed:
Update transmission.jl

Deleted unused EXPANSION_LINES and EXPANSION_LINES_ASYMMETRIC

sambuddhac committed:
Update transmission.jl

sambuddhac committed:
Fixed the multistage asymmetric flow bug in file and added a new example case

sambuddhac committed:
Introduced try catch blocks to handle situations where the asymmetrical flag is turned on but the corresponding columns are not provided in the Networks.csv

qluo0320github commented:
I believe "asymmetrical_trans_flow_limit" should be used in settings,
instead of AsymmetricalTransFlowLimit.

On Thu, Feb 27, 2025, 11:37 Chakrabarti, Sambuddha (Sam) <
***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/load_inputs/load_network_data.jl
> <https://github.com/GenXProject/GenX.jl/pull/789#discussion_r1973957500>:
>
> > +    if setup["asymmetrical_trans_flow_limit"] == 1
> +        L_asym = length(filtered_vector(network_var, :Asymmetrical, 1, :Network_Lines)) #Number of asymmetrical lines
> +    end
>
> @lbonaldo <https://github.com/lbonaldo> I fixed this with the latest push
> and tested also. In order to review, for instance, you could take any of
> the usual example cases (i.e. not the asymmetrical cases) and in the
> settings file, put the AsymmetricalTransFlowLimit switch and set it to 0
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/GenXProject/GenX.jl/pull/789#discussion_r1973957500>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALBIOTT77JINVQFFQJPLHND2R45M7AVCNFSM6AAAAABRXAC3KKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMNBYGQYTEMBWGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

sambuddhac committed:
Changed to uppercase for SYMMETRIC_LINE_INDEX and ASYMMETRIC_LINE_INDEX

sambuddhac committed:
Introduced the intersections for symmetric loss lines and symmetric expansion lines in network load file to speed up computation

sambuddhac committed:
Included the transmission losses in the write outputs

sambuddhac committed:
Modified the precompile case for incorporating asymmetrical bidirectional flows and fixed the loss writing

sambuddhac commented:
@lbonaldo this is now fully ready (except that one tiny part where we might want to finalize the set of input example cases; what I can also do, is to change the Neg in all of them and re-number the cases, I can do that tomorrow). Also, I rebased with `develop`, but, there's a very tiny little update on the changelog that was giving me a hard time. Maybe, it can be more easily fixed when you finally merge it on `develop`. Also, @qluo0320github , this might be a good time to do another round of testing and see if there's any further issues. If yes, please lmk. If everything looks good, this can be merged totally.

sambuddhac committed:
placed the units at the end of the key with Neg and code cleanup

sambuddhac commented:
@lbonaldo , @qluo0320github , @JesseJenkins , done with everything on this PR. It should now be fully polished and ready to be merged. Please take a look and see if there's any more suggestions.

sambuddhac committed:
Modified DDP file to have transmission builds across stages; Added example cases to test multistage asymmetric builds

sambuddhac committed:
Modifications to ddp.jl for incorporating multistage asymmetric line flow

sambuddhac committed:
Add warning for asymmetric flow issues

Add warning for potential issues with asymmetric flow in multistage capacity expansion.

sambuddhac committed:
Added try catch block for error handling for asymmetric multistage

sambuddhac committed:
Modified write_nw_expansion.jl to print Positive and Negative Transcap

sambuddhac committed:
Modified to used asymmetric expansion lines for correct indexing

sambuddhac committed:
Made the columns of equal length in the output network writing and corrected indexing

lbonaldo committed:
Code clean up after rebase

lbonaldo committed:
Fix latex symbols in docstrings

lbonaldo committed:
write new capacity for asym lines to output files

lbonaldo committed:
Fix backward compatibility with symmetric lines and multistage

lbonaldo committed:
Update tracking of tx capacity in multistage expansion

sambuddhac committed:
Merge branch 'develop' into asymmetric_line_flow

sambuddhac reviewed (commented):
Reviewed and approved !!!

lbonaldo commented on precompile/case/settings/genx_settings.yml:
```diff
@@ -6,4 +6,5 @@ MinCapReq: 1
 UCommit: 2
-OverwriteResults: 1
\ No newline at end of file
+OverwriteResults: 1
+AsymmetricalTransFlowLimit: 1
```
I probably wouldn't modify the precompile case. Do we have a specific reason for this change?

sambuddhac replied:
Thanks for the feedback, Luca. I think the reason I was doing it initially, is because I wasn't sure if the changes would impact the precompile. But, we can surely have the original precompile.

lbonaldo commented on src/write_outputs/transmission/write_nw_expansion.jl:
```diff
@@ -1,22 +1,62 @@
+    L = inputs["L"]
+
+    NetworkExpansion = setup["NetworkExpansion"]
+    MultiStage = setup["MultiStage"]
```
Is this line being used?

lbonaldo commented on src/write_outputs/transmission/write_transmission_flows.jl:
```diff
@@ -4,7 +4,10 @@ function write_transmission_flows(path::AbstractString,
     # Transmission related values
     T = inputs["T"]     # Number of time steps (hours)
-    L = inputs["L"]     # Number of transmission lines
+    L_sym = inputs["L_sym"] # Number of transmission lines with symmetrical bidirectional flow
+    L_asym = inputs["L_asym"] #Default number of asymmetrical lines
+    # Number of lines in the network
+    L = L_sym + L_asym
```
Are `L_sym` and `L_asym` being used in other places other than at this line? If not, we could keep using `L`.

lbonaldo commented on src/configure_settings/configure_settings.jl:
```diff
@@ -37,7 +37,9 @@ function default_settings()
         "PoliciesFolder" => "policies",
-        "ObjScale" => 1)
+        "ObjScale" => 1,
+        "AsymmetricalTransFlowLimit" => 0
```
We need to update the documentation with this new settings key.

lbonaldo reviewed (changes requested):
Hi @sambuddhac, thank you for your recent changes to this PR. I’m doing a more comprehensive review, but in the meantime, I added a few more comments.
Do we want to keep all the new example cases? Right now we have 8 example cases for single stage, and 4 for multi-stage.
